src/utils/work/seekWorkResponse.js

```diff
@@ -146,30 +134,43 @@ const dispatchWorkRequest = async (
   });
 
   const responsePromise = new Promise((resolve, reject) => {
-    io.on(`WorkResponse-${ETag}`, (res) => {
-      const { response } = res;
-
-      if (response.error) {
-        const { errorCode, userMessage } = response;
-        console.error(errorCode, userMessage);
-
-        return reject(
-          new WorkResponseError(errorCode, userMessage, request),
-        );
+    io.on(`WorkResponse-${ETag}`, async (res) => {
+      // If type is object, then we received a notification with a signedUrl
+      // now we need to fetch the actual result from s3
+      if (typeof res === 'object') {
+        const { response } = res;
+
+        if (response.error) {
+          const { errorCode, userMessage } = response;
+          console.error(errorCode, userMessage);
+
+          return reject(
+            new WorkResponseError(errorCode, userMessage, request),
+          );
+        }
+
+        return resolve({ signedUrl: response.signedUrl });
       }
 
-      // If no error, the response should be ready on S3.
-      // In this case, return true
-      return resolve();
+      // If type isn't object, then we have the actual work result,
+      // no further downloads are necessary, we just need to decompress and return it
+      const decompressedData = await decompressUint8Array(Uint8Array.from(Buffer.from(res, 'base64')));
+      return resolve({ data: parseResult(decompressedData) });
     });
   });
 
-  const result = Promise.race([timeoutPromise, responsePromise]);
+  // TODO test what happens when api throws an error here
+  await fetchAPI(
+    `/v2/workRequest/${experimentId}/${ETag}`,
+    {
+      method: 'POST',
+      headers: { 'Content-Type': 'application/json' },
+      body: JSON.stringify(request),
+    },
+  );
```

Comment:
I joined the roles of the socket WorkRequest-v2 and `workResults/` http endpoint into one single endpoint `workRequest`.

This endpoint checks if the result is stored
  - If it is, it returns over the `WorkResults-{ETag}` the signedUrl.
  - If it isn't, it submits the work request

Reply:
I believe this is better than our previous options.

- the `WorkRequest-v2` didn't need to be a socket and made the UI a bit more complicated handling different ways of submitting work requests. I would consider using a socket (redis adapter) though if we somehow used it to connect directly to the worker to avoid the api as a middleman, but the danger there is that users can send whatever crap they want directly to the R worker. Also I tried this out in staging and didn't notice a significant improvement (the api and worker are in the same region, so added delay is relatively insignificant